First working version of audio system

MagrosThornrime opened the pull request:
The audio system is now integrated with ECS. It allows for playing multiple sounds at once. When a sound stops - because it reached the end or the users stopped it - the system automatically removes it. The user can pause and continue all sounds. 

~~I've written 3 simple tests:~~
~~1. When you click enter, play a short sound, which is deleted after some time~~
~~2. Play a song that is looped. You can pause it, continue, stop (and then play it another time) as you wish~~
~~3. For some time, play a wind sound. Its source rotates around the listener, while changing the distance from them. It's meant to test the spatial audio.~~

Now we have only one example, related to spatial audio. But it also uses our renderer to show where the sound's source really is.

The system supports multiple file formats. I tested .wav and .mp3 files. It also supports dynamic loading of files, so we can load just the files we need - and spare some memory.

I wrote some documentation using comments, but I need to write unit tests yet.

Twarug commented on include/audio/AudioSource.h:
```diff
@@ -0,0 +1,74 @@
+#pragma once
+#include <string>
+
+#include <AL/al.h>
```
I don't think that a library (implementation detail) should be part of Archimedes API.
Includes to other libs should be done only in cpp files or header files that are not part of an our API.

Chris-plusplus replied:
However, AL/al.h defines typedefs used later in the file.
C++ modules could fix this problem, but we all know hoe that would work.

Twarug replied:
These types also shouldn't be used in our API.

MagrosThornrime replied:
Most of the types are just aliases, e.g. ALfloat is just a float. I can replace them with C++ equivalents and find out if some aren't convertible (like OpenAL enums)

Twarug replied:
In case of floats use our math types (eg. `f32`).
For enums create your own, and in implementation map between them as needed. (eg. gfx::GraphicsFormat)

MagrosThornrime replied:
As I expected, there were some types that I couldn't simply replace. You need to access the sound device (your soundcard) and it's stored as a struct. Also, you need to free the struct with special OpenAL call after using. So i can't remove/replace these parameters from the API

MagrosThornrime replied:
Fixed, thanks to @Chris-plusplus 
In the API, those structs are casted to char*, and when used, they are casted to OpenAL types

Twarug commented on include/audio/Calls.hpp:
```diff
@@ -0,0 +1,137 @@
+    /// @param params wrapped function's parameters
+    template<typename AlFunction, typename... Params>
+    requires voidReturn<AlFunction, Params...>
+    auto alCall(AlFunction function, Params... params){
```
Forwarded params should always be a rvalue reference (&&)

MagrosThornrime replied:
This file is somewhat cursed and needs a bigger rework. But I will look into this

MagrosThornrime replied:
replaced with:
```cpp
auto alCall(AlFunction function, Params&&... params) {
```

Twarug commented on include/audio/Calls.hpp:
```diff
@@ -0,0 +1,137 @@
+    /// @return return value of wrapped function
+    template<typename AlFunction, typename... Params>
+    requires normalReturn<AlFunction, Params...>
+    auto alCall(AlFunction function, Params... params){
```
Forwarded params should always be a rvalue reference (&&)

MagrosThornrime replied:
Replaced with:
```cpp
auto alCall(AlFunction function, Params&&... params) {
```

Twarug commented on include/audio/Calls.hpp:
```diff
@@ -0,0 +1,137 @@
+    /// @param params wrapped function's parameters
+    template<typename AlcFunction, typename... Params>
+    requires voidReturn<AlcFunction, Params...>
+    auto alcCall(AlcFunction function, ALCdevice* device, Params... params){
```
Forwarded params should always be a rvalue reference (&&)

MagrosThornrime replied:
replaced with:
```cpp
auto alcCall(AlcFunction function, ALCdevice* device, Params&&... params) {
```

Twarug commented on include/audio/Calls.hpp:
```diff
@@ -0,0 +1,137 @@
+    template<typename AlcFunction, typename... Params>
+    requires normalReturn<AlcFunction, Params...>
+    auto alcCall(AlcFunction function, ALCdevice* device, Params... params){
+        auto returnValue = function(std::forward<Params>(params)...);
```
Forwarded params should always be a rvalue reference (&&)

MagrosThornrime replied:
Replaced with:
```cpp
template<typename AlcFunction, typename... Params>
requires normalReturn<AlcFunction, Params...>
auto alcCall(AlcFunction function, ALCdevice* device, Params&&... params) {
	auto returnValue = function(std::forward<Params>(params)...);
```

Twarug commented on include/audio/Clip.h:
```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#include <AL/al.h>
```
Library import in API

MagrosThornrime replied:
removed

Twarug commented on include/audio/Listener.h:
```diff
@@ -0,0 +1,35 @@
+#pragma once
+#include <AL/al.h>
```
Lib import in API

MagrosThornrime replied:
removed

Twarug commented on include/audio/SoundDevice.h:
```diff
@@ -0,0 +1,21 @@
+#pragma once
+#include <AL/alc.h>
```
API import in API

MagrosThornrime replied:
It's hard to avoid because SoundDevices uses OpenAL structs as parameters. And they can't be removed, as you need to free these resources after audio system shutdown

Twarug replied:
In API, u can make some sort of type erasure using `void*` or something like this

MagrosThornrime replied:
I missed the comment, @Chris-plusplus mentioned it when we talked on discord. So thanks to you two

Twarug commented on include/audio/SourcePlayer.h:
```diff
@@ -0,0 +1,115 @@
+#pragma once
+
+#include <AL/al.h>
```
Lib import in API

MagrosThornrime replied:
removed

Twarug commented on include/audio/AudioSource.h:
```diff
@@ -0,0 +1,74 @@
+
+	/// @brief ECS component used to store all data about a sound source on the scene.
+	/// It also allows to specify what the AudioManager should do with it.
+	struct AudioSource{
```
Should be named `AudioSourceComponent` if it is one

MagrosThornrime replied:
It was, but I thought that it would be better to name it like AudioSource objects from Unity. Because it's the object directly by the user (contrary to e.g SourcePlayer which is just a slot for OpenAL source and is responsible for playing the audio).

MagrosThornrime replied:
Probably I will do it, cause I've seen that you also used "Component" as a suffix to all components' names. So it might become our convention for naming all the Components.

Twarug commented on src/audio/AudioManager.cpp:
```diff
@@ -0,0 +1,114 @@
+	while (_isListening) {
+		_updateListener();
+		{
+			auto lock = std::lock_guard(_ecsMutex);
```
Locking entire ECS domain isn't good idea.

Twarug commented on include/audio/AudioSource.h:
I thin better approach would be not having in component any state, that is modified by Audio Manager, only settings, modulations of an Audio.
Things like state could be tracked internally by manager.
Allowing removal of ECS Domain lock, which isn't a good idea in the first place.
Requests to play/stop ect. Source form playing could be done by adding tag components (eg. `StopAudioSourceComponet`), witch stops clip from current entity on next Audio System update.

MagrosThornrime replied:
I don't like the idea of tag components, but I haven't seen them in practice. I'm rather thinking about storing references/pointers to AudioSources in the AudioManager. The problem was: "what if somebody deletes the AudioSource?". Because then we would have either a nullptr or potentially an illegal pointer (leading to segfaults and alike).

But then I realized that I can just use the "listener" design pattern - each AudioSource will inform the AudioManager about their destruction. Then the manager will drop it.

Chris-plusplus replied:
Using flag-components could have its benefits. For example: instead of AudioSources manually informing AudioManager by calling its method, the manager would simply have a list of AudioSources that have been stopped with appropriate flag-component.

Twarug commented on src/audio/AudioManager.cpp:
```diff
@@ -0,0 +1,114 @@
+		_updateListener();
+		{
+			auto lock = std::lock_guard(_ecsMutex);
+			for (auto&& [source] : _domain->view<AudioSource>().components()) {
```
Audio system update, should iterate over const Components, as they should not contain state, only settings.

Twarug commented on src/audio/AudioManager.cpp:
```diff
@@ -0,0 +1,114 @@
+	Logger::info("Audio system: removed Source with index {}", std::to_string(index));
+}
+
+void AudioManager::_pauseSource(AudioSource& source) {
```
Should be const reference

Twarug commented on src/audio/AudioManager.cpp:
```diff
@@ -0,0 +1,114 @@
+	_sources[source.sourceIndex].pausePlaying();
+}
+
+void AudioManager::_runSource(AudioSource& source) {
```
Should be const reference

Twarug commented on src/audio/AudioManager.cpp:
```diff
@@ -0,0 +1,114 @@
+	_sources[source.sourceIndex].run(source);
+}
+
+bool AudioManager::_stopSource(AudioSource& source) {
```
Should be const reference

Twarug commented on src/audio/AudioManager.cpp:
```diff
@@ -0,0 +1,114 @@
+	return _sources[source.sourceIndex].stopPlaying();
+}
+
+void AudioManager::_updateSource(AudioSource& source) {
```
Should be const reference

Twarug commented on src/audio/AudioManager.cpp:
```diff
@@ -0,0 +1,114 @@
+	Logger::info("Audio system: assigned Source with index {}", std::to_string(source.sourceIndex));
+}
+
+void AudioManager::_removeSource(AudioSource& source) {
```
In consequence of above changes, Should be const reference

Twarug commented on src/audio/AudioManager.cpp:
```diff
@@ -0,0 +1,114 @@
+	_isListening = false;
+}
+
+void AudioManager::_assignSource(AudioSource& source) {
```
In consequence of above changes, Should be const reference

Twarug commented on src/net/Socket.cpp:
```diff
@@ -1,3 +1,5 @@
+#include <stdexcept>
```
Why?

MagrosThornrime replied:
Probably just a typo. Or I was trying to help Krystian with networking system, I don't remember. Removing it.

Twarug reviewed (commented):
Over all system looks good, but there are somo questionable design choices

P.S. In few files I spotted lack of space after namespace/class name and `{`

MagrosThornrime commented on .clang-format:
```diff
@@ -63,7 +63,7 @@ AlignConsecutiveShortCaseStatements:
 AllowAllArgumentsOnNextLine: false
 AllowAllParametersOfDeclarationOnNextLine: false
-AllowBreakBeforeNoexceptSpecifier: OnlyWithParen
+# AllowBreakBeforeNoexceptSpecifier: OnlyWithParen
```
Maybe clion hinted it before and I changed it, I don't remember. Now fixed.

MagrosThornrime commented on archimedes_bin/main.cpp:
Good idea. In which directory should I put the examples? I think they shouldn't be in the same place with tests, because they are more interactive (like our demo)

Twarug replied:
You can rebase/merge upsteam and look at way I implemented example for NvrhiRenderer

MagrosThornrime replied:
Done. We have now an example that uses AudioSystem and your renderer simultaneously

MagrosThornrime commented:
Seeing that you all are discussing in english, I will translate the main comment cause it doesn't fit in

MagrosThornrime commented:
@Twarug I'm not sure if I need also to add rvalue references to concepts, like here:
```cpp
template<typename Function, typename... Params>
/// @brief Check if function returns void
concept voidReturn = requires(Function function, Params... params) {
	{ function(std::forward<Params>(params)...) } -> std::same_as<void>;
};
```
Because we still forward the parameters

MagrosThornrime commented:
Also, I need to make Listener a public variable (maybe a Component even), use std::filesystem instead of hardcoding file paths and make all AudioExceptions' messages more readable. Before closing the request. Because these stuff is just obvious.

I've thought of upgrading the Calls.hpp file but I don't feel that it will be an easy task. Using deduction guides, I could fix std::sourcelocation feature. After some time spent at debugging the audio system, it shows that it would be REALLY helpful.

MagrosThornrime commented:
Don't go by the latest commit, it broke SO MAAAANY things... I'm trying to use tag components for storing AudioSourceComponents' state. Will fix it later.

MagrosThornrime commented:
I rewrote the AudioManager to avoid locking the entire domain and removed the states from AudioSourceComponents. Also, I fixed some weird design errors, like hardcoding paths to audio files.

Calls.hpp contains now two macros that allow us to check where exactly OpenAL threw an error. And we can update almost all important parameters of audio sources and the listener. We have a working example that uses the new version of the Renderer.

Twarug commented on archimedes_bin/examples/audio/LegacyAudioTests.h:
Those test files should be an Application

MagrosThornrime replied:
It's impossible until we have a working input manager. They are interactive

MagrosThornrime replied:
Also, SpatialAduoTestApp is an Application

MagrosThornrime replied:
Ok, Krystian somehow got the input from the system, I will try to rewrite the examples

MagrosThornrime replied:
Done

MagrosThornrime commented on .gitignore:
I used venv for conan installation. I don't know how the "thirdparty" directory appeared, but it's also in my project directory.

MagrosThornrime replied:
Removed

MagrosThornrime commented on CMakeLists.txt:
```diff
@@ -15,7 +15,7 @@ include("cmake/msvc.cmake")
 
 # uncomment below to force conan reinstallation
-#set(ARCHIMEDES_FORCE_CONAN_INSTALL TRUE)
+set(ARCHIMEDES_FORCE_CONAN_INSTALL TRUE)
```
Done, will appear in the next commit.

MagrosThornrime commented on .clang-format:
I have no idea, I will upload the version from the master in the next commit.

MagrosThornrime commented on archimedes_bin/main.cpp:
I used this to test the audio system. I will paste the code from the master in the next commit.

MagrosThornrime commented on include/audio/ListenerComponent.h:
Sounds right, I will update it

MagrosThornrime replied:
Updated

MagrosThornrime commented:
All stuff you pointed out fixed, thanks for help!
